### PR TITLE
Do not delete bootstrap secret until machine object is deleted

### PR DESCRIPTION
**Pull request by jakefhyde**
## Problem

The RKEBootstrap data secret is deleted after the associated CAPI Machine object transitions into the Provisioned phase. This prevents upstream infrastructure providers from working properly, since the system-agent will fail to connect if the bootstrap secret does not exist.

## Solution

Do not delete the bootstrap secret until the machine is deleted.

I also considered waiting until the machine phase was `Running` to delete the bootstrap secret, however this implementation keeps us in parity with upstream, and I feel strongly that if we are going to make a deviation it needs to be purposeful and documented.

### Manual Testing

Tested manually with both v2prov docker machine based infrastructure providers, and upstream digital ocean provider.

### Automated Testing

N/A

## QA Testing Considerations

N/A. QA will not be able to test that this works with upstream infrastructure providers, but should note that the bootstrap secrets will be viewable for the related machine objects now. This can be tested by provisioning and rke2/k3s node driver cluster, and navigating to Secrets in the UI, and inspecting that a corresponding secret in the `fleet-default` namespace matching the name of the bootstrap object's `DataSecretName` field is present.

### Regressions Considerations

N/A

### CODE REVIEW

File: pkg/controllers/provisioningv2/rke2/bootstrap/controller.go [hunk 1]
```diff
@@ -223,11 +223,14 @@ func (h *handler) getEnvVar(bootstrap *rkev1.RKEBootstrap, capiCluster *capi.Clu
...
+// until the corresponding Capi Machine object's Machine Phase is at least "Running", which indicates that the machine
+// "has become a Kubernetes Node in a Ready state".
 func (h *handler) assignBootStrapSecret(machine *capi.Machine, bootstrap *rkev1.RKEBootstrap, capiCluster *capi.Cluster) (*corev1.Secret, []runtime.Object, error) {
-	if capi.MachinePhase(machine.Status.Phase) != capi.MachinePhasePending &&
-		capi.MachinePhase(machine.Status.Phase) != capi.MachinePhaseDeleting &&
-		capi.MachinePhase(machine.Status.Phase) != capi.MachinePhaseFailed &&
-		capi.MachinePhase(machine.Status.Phase) != capi.MachinePhaseProvisioning {
+	if capi.MachinePhase(machine.Status.Phase) == capi.MachinePhaseDeleting &&
+		capi.MachinePhase(machine.Status.Phase) == capi.MachinePhaseDeleted &&
+		capi.MachinePhase(machine.Status.Phase) == capi.MachinePhaseFailed {
```
Comment (Oats87) [inline pkg/controllers/provisioningv2/rke2/bootstrap/controller.go:233]:
```suggestion
	if capi.MachinePhase(machine.Status.Phase) == capi.MachinePhaseDeleting ||
		capi.MachinePhase(machine.Status.Phase) == capi.MachinePhaseDeleted ||
		capi.MachinePhase(machine.Status.Phase) == capi.MachinePhaseFailed {
```

**Review by Oats87 — CHANGES_REQUESTED** [review]
I'm assuming this is because CAPI intends machine phase = `provisioned` to mean the infra is running but not necessarily bootstrapped, whereas our rke-machine provider sees it as bootstrapped?

I would be curious to see what happens if a user re-runs the cloud-init (install script) after a node is successfully bootstrapped, because before the script would fail and the token that was included in the cloud-init data was invalid.

File: pkg/controllers/provisioningv2/rke2/bootstrap/controller.go (hunk 1 above)
Comment (jakefhyde) [reply pkg/controllers/provisioningv2/rke2/bootstrap/controller.go]: You ever go through the `git commit --amend && git push -u origin --force` and realize you forgot to `git add`? Feel like that happens to me at least once a week. Anyways, should be correct now.

**Comment by jakefhyde** [comment]
> I'm assuming this is because CAPI intends machine phase = `provisioned` to mean the infra is running but not necessarily bootstrapped, whereas our rke-machine provider sees it as bootstrapped?
> 
> I would be curious to see what happens if a user re-runs the cloud-init (install script) after a node is successfully bootstrapped, because before the script would fail and the token that was included in the cloud-init data was invalid.

Your assumption is correct, these are the expected machine phases:
```
	// MachinePhasePending is the first state a Machine is assigned by
	// Cluster API Machine controller after being created.
	MachinePhasePending = MachinePhase("Pending")

	// MachinePhaseProvisioning is the state when the
	// Machine infrastructure is being created.
	MachinePhaseProvisioning = MachinePhase("Provisioning")

	// MachinePhaseProvisioned is the state when its
	// infrastructure has been created and configured.
	MachinePhaseProvisioned = MachinePhase("Provisioned")

	// MachinePhaseRunning is the Machine state when it has
	// become a Kubernetes Node in a Ready state.
	MachinePhaseRunning = MachinePhase("Running")

	// MachinePhaseDeleting is the Machine state when a delete
	// request has been sent to the API Server,
	// but its infrastructure has not yet been fully deleted.
	MachinePhaseDeleting = MachinePhase("Deleting")

	// MachinePhaseDeleted is the Machine state when the object
	// and the related infrastructure is deleted and
	// ready to be garbage collected by the API Server.
	MachinePhaseDeleted = MachinePhase("Deleted")

	// MachinePhaseFailed is the Machine state when the system
	// might require user intervention.
	MachinePhaseFailed = MachinePhase("Failed")

	// MachinePhaseUnknown is returned if the Machine state cannot be determined.
	MachinePhaseUnknown = MachinePhase("Unknown")
```

Note the state transition from `Provisioning` to `Provisioned`, which doesn't encapsulate any information regarding the bootstrapping process. I think the difference is that the infrastructure being ready does not account for cloud-init having completed successfully, whereas it may for machine (or the time that it takes for the machine to be created and the status to trickle out of the machine-provision pod back onto the object is long enough that we don't see this). Glancing at the `v3/connect/agent` endpoint, I'm assuming the token was invalidated before it could connect, but if the token expiring is desired/expected behavior we can delete it once the phase transitions to `Running`.

**Review by Oats87 — APPROVED** [review]

**Review by jiaqiluo — APPROVED** [review]

**Comment by Oats87** [comment]
Verified that the bootstrap secret is no longer deleted when the machine transitions to Provisioned/Running.

On a v2.7.5 setup:

```
> kubectl get secrets -n fleet-default
NAME                                                           TYPE                                          DATA   AGE
import-token-test                                              fleet.cattle.io/cluster-registration-values   1      14s
import-token-test-edeea398-d34f-4abe-ac77-31ea2b8557ba-token   kubernetes.io/service-account-token           3      16s
test-bootstrap-template-8gdjp-machine-plan                     rke.cattle.io/machine-plan                    13     5m
test-bootstrap-template-8gdjp-machine-plan-token-z7z4k         kubernetes.io/service-account-token           3      4m19s
test-kubeconfig                                                Opaque                                        2      5m
test-pool1-5a13a63e-2ms76                                      Opaque                                        2      4m41s
test-pool1-5a13a63e-2ms76-machine-state                        rke.cattle.io/machine-state                   1      4m49s
test-rke-state                                                 rke.cattle.io/cluster-state                   2      5m
> kubectl get machines -n fleet-default
NAME                          CLUSTER   NODENAME                    PROVIDERID                         PHASE     AGE     VERSION
test-pool1-64d58bbd99-ngjhm   test      test-pool1-5a13a63e-2ms76   rke2://test-pool1-5a13a63e-2ms76   Running   5m15s   
> 
```

On a v2.7-head setup:
```
> kubectl get secrets -n fleet-default
NAME                                                           TYPE                                          DATA   AGE
import-token-test                                              fleet.cattle.io/cluster-registration-values   1      71s
import-token-test-2e2128ef-77ff-41bb-8ddf-5aaa4c0df599-token   kubernetes.io/service-account-token           3      73s
test-bootstrap-template-k4hml-machine-bootstrap                rke.cattle.io/bootstrap                       1      5m40s
test-bootstrap-template-k4hml-machine-bootstrap-token-65sdb    kubernetes.io/service-account-token           3      5m40s
test-bootstrap-template-k4hml-machine-plan                     rke.cattle.io/machine-plan                    13     5m40s
test-bootstrap-template-k4hml-machine-plan-token-d7v8n         kubernetes.io/service-account-token           3      4m53s
test-kubeconfig                                                Opaque                                        2      5m40s
test-pool1-5a13a63e-x5gxv                                      Opaque                                        2      5m17s
test-pool1-5a13a63e-x5gxv-machine-state                        rke.cattle.io/machine-state                   1      5m30s
test-rke-state                                                 rke.cattle.io/cluster-state                   2      5m40s
> kubectl get machines -n fleet-default
NAME                                CLUSTER   NODENAME                    PROVIDERID                         PHASE     AGE     VERSION
test-pool1-578d46948fxs98p9-d98pj   test      test-pool1-5a13a63e-x5gxv   rke2://test-pool1-5a13a63e-x5gxv   Running   6m12s  
```

Note the existence of the `test-bootstrap-template-k4hml-machine-bootstrap` and corresponding `test-bootstrap-template-k4hml-machine-bootstrap-token-65sdb` within the v2.7-head setup.

**Pull request merged**